Added Support for SPL Token-2022

luffy-orf opened the pull request:
## Description - 
Added comprehensive support for SPL Token-2022 program, enabling the handling of advanced token features like transfer fees, interest-bearing tokens, and other extensions.

### Solves -> https://github.com/solana-foundation/kora/issues/18
## Changes
- Added Token2022 program ID and account state handling
- Implemented support for Token2022 extensions:
  - Transfer fees
  - Interest-bearing tokens
  - Non-transferable tokens
  - CPI guard
  - Confidential transfers
- Enhanced validation logic to handle Token2022 specific features
- Added integration tests for Token2022 operations

## Testing
- Unit tests for Token2022 program ID verification
- Integration tests for token operations
- Tests for ATA derivation and transfer instructions
- Coverage for extension-specific functionality

. 
 This pull request was created for https://app.gib.work/bounties/28402620-7287-4cb2-9db6-10dbe2307228 in an attempt to solve a bounty #18 . Payment for the bounty is immediately sent to the contributor after merge.

luffy-orf commented:
@gitteri @pkxro Please review the changes and let me know if any improvements or modifications are needed

gitteri commented on crates/lib/src/token/tests.rs:
```diff
@@ -80,4 +183,312 @@ mod tests {
+
+        // Test unpacking with our implementation
+        let program = TokenProgram::new(TokenType::Token2022);
+        let unpacked = program.unpack_token_account(&buffer).unwrap();
```
Please add a test that takes a real token deployed to Solana (like PYUSD) and unpack it using this function and verify its details. You should hard code the token bytes in the test so it does not need a network connection

gitteri commented on crates/lib/src/transaction/validator.rs:
```diff
@@ -239,63 +247,149 @@ impl TransactionValidator {
-            continue;
-        }
+    // Check both SPL and Token2022 transfers
+    for token_type in [TokenType::Spl, TokenType::Token2022] {
```
you shouldn't need to iterate twice. You can check the instruction program id against these two program ids to see which one to use on line 263.

gitteri commented on crates/lib/src/transaction/validator.rs:
```diff
@@ -239,63 +247,149 @@ impl TransactionValidator {
+                        }
+                    }
+
+                    // Check confidential transfers
+                    if token2022_account.confidential_transfer.is_some() {
+                        return Err(KoraError::InvalidTransaction(
+                            "Confidential transfers not supported".to_string(),
+                        ));
+                    }
+
+                    // Calculate actual transfer amount considering all extensions
+                    let mut actual_amount = amount;
+
+                    // Apply transfer fee if present
+                    if let Some(fee_config) = &token2022_account.transfer_fee {
+                        let basis_points = u16::from_le_bytes(
+                            fee_config.newer_transfer_fee.transfer_fee_basis_points.0,
+                        );
+                        let fee = std::cmp::min(
+                            (amount as u128 * basis_points as u128 / 10000) as u64,
+                            u64::from(fee_config.newer_transfer_fee.maximum_fee),
+                        );
+                        actual_amount = actual_amount.saturating_sub(fee);
+                    }
+
+                    // Apply interest if present
+                    if let Some(interest_config) = &token2022_account.interest_bearing {
+                        // Calculate accrued interest
+                        let current_ts = std::time::SystemTime::now()
+                            .duration_since(std::time::UNIX_EPOCH)
+                            .unwrap()
+                            .as_secs();
+                        let interest_rate = u16::from_le_bytes(interest_config.current_rate.0);
+                        let last_update: i64 = interest_config.last_update_timestamp.into();
+                        let time_delta = current_ts.saturating_sub(last_update as u64);
+                        let interest =
+                            (actual_amount as u128 * interest_rate as u128 * time_delta as u128)
+                                / (10000 * 365 * 24 * 60 * 60) as u128;
+                        actual_amount = actual_amount.saturating_add(interest as u64);
+                    }
```
Please move this to a separate function for readability and improved testing

gitteri commented on crates/lib/src/transaction/validator.rs:
```diff
@@ -239,63 +247,149 @@ impl TransactionValidator {
+                        actual_amount = actual_amount.saturating_sub(fee);
+                    }
+
+                    // Apply interest if present
+                    if let Some(interest_config) = &token2022_account.interest_bearing {
```
I think its fair to assume the amount requested will be in raw amounts and not the amount including interest. calculate_token_value_in_lamports should use a price source that is based on the raw amount and not based on the compounding interest

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -26,4 +27,55 @@ async fn test_token_operations() {
+    assert_eq!(create_ata_ix.program_id, spl_associated_token_account::id());
+}
+
+#[test]
+fn test_token2022_program_id() {
+    let program = TokenProgram::new(TokenType::Token2022);
+    assert_eq!(program.program_id(), spl_token_2022::id());
```
why is this in the integration tests file? move it to the token/tests.rs file if this is needed

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -26,4 +27,55 @@ async fn test_token_operations() {
+}
+
+#[tokio::test]
+async fn test_token2022_operations() {
```
It'd be better for these integration tests to use Kora and a known devnet t22 token like PYUSD to validate an e2e transfer operation works.

luffy-orf commented:
@gitteri Please review the changes

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -1,29 +1,172 @@
+use std::str::FromStr;
+
+// PYUSD token mint on devnet
+const PYUSD_MINT: &str = "PYUSD6FiXyYotH6a4LJDP8mjQT7eJRJ4mkd5ZUj94UF";
```
PYUSD is actually at CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM
You can use paxos' faucet to get some if you need

gitteri reviewed (changes requested):
Great progress! Left comments to address

gitteri commented on crates/lib/src/token/mod.rs:
```diff
@@ -2,11 +2,13 @@ use crate::error::KoraError;
-mod implementation;
 mod interface;
+mod token;
+mod token22;
```
you have to add this to actually run the new tests. There are a few that are currently failing
```
#[cfg(test)]
mod tests;
```

luffy-orf commented:
@gitteri  Tests are running fine now. Please review the changes.

gitteri commented:
token_integration_tests.rs needs some updates and currently doesn't pass. @luffy-orf please review

luffy-orf commented:
> token_integration_tests.rs needs some updates and currently doesn't pass. @luffy-orf please review

@gitteri  I have fixed the tests

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -66,36 +102,57 @@ async fn test_pyusd_token_e2e_with_kora() {
         token_program.get_associated_token_address(&destination_wallet.pubkey(), &pyusd_mint);
 
-    // Create a transfer instruction
-    let transfer_ix = token_program
```
I'm worried using the helper is masking a bug

luffy-orf replied:
fixed it by directly using the Token2022 program ID instead of the helper function.

gitteri commented:
@luffy-orf why can't you use the token interface directly for this? you shouldn't need a helper function in the integration tests.

luffy-orf commented:
> @luffy-orf why can't you use the token interface directly for this? you shouldn't need a helper function in the integration tests.

@gitteri  Totally right — I was overthinking it and added a helper that wasn’t needed. Switched to Token2022Program::new() and now just using TokenInterface methods directly. Appreciate you catching that!

luffy-orf commented:
@gitteri  — just checking, are there any other changes required in the PR?

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -1,29 +1,194 @@
-use kora_lib::token::{TokenProgram, TokenType, TokenInterface};
-use solana_client::rpc_client::RpcClient;
-use solana_sdk::pubkey::Pubkey;
+use kora_lib::{
```
nit: there are a bunch of unused

gitteri commented on tests/token_integration_tests.rs:
```diff
@@ -1,29 +1,194 @@
+    let recent_blockhash = rpc_client.get_latest_blockhash().await.unwrap();
+
+    // Create a transaction to create the token account
+    let create_account_tx = Transaction::new_signed_with_payer(
```
nit: unused
